hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -624,6 +624,37 @@ def do_draw(self, data):
                 # As long as we consume data, we'll eventually pass or raise.
                 # But if we don't this could be an infinite loop.
                 assume(data.index > start_index)
+
+        # Special-case for SampledFrom strategy.  This is special even among
+        # strategies with a known and small(ish) list of possible elements,
+        # because it implements drawing directly as choice of an index.
+        # That means we can use the "make your own luck" trick explained in
+        # the shrinking guide to find even very rare allowed values.
+        from hypothesis.searchstrategy.lazy import unwrap_strategies
+        from hypothesis.searchstrategy.misc import SampledFromStrategy
+
+        strategy = unwrap_strategies(self.filtered_strategy)
+        if isinstance(strategy, SampledFromStrategy):
```

Comment:
I am reasonably keen on not special casing this in `filter` but instead in `SampledFromStrategy` and handling the lazy filter delegation as I described earlier. The reason is that this isn't a problem specific to `SampledFrom` and we may end up wanting to special case it elsewhere too. The obvious common other examples where it's an issue are `integers(m, n)` (almost certainly worth special casing, and even more so on the parallel `unique=True` branch), and `booleans` (probably not worth special casing). It's plausibly also worth us thinking about special casing `one_of` in filter.

I'm not proposing you *do* any of these special cases in this branch (indeed I would strongly discourage it), but I think at least allowing for the possibility of them is important.

Reply:
Earlier in https://github.com/HypothesisWorks/hypothesis/pull/1890#discussion_r268412292 I proposed the possibility of a per-strategy hook that `FilteredStrategy.do_draw` could detect and call, after rejection-sampling fails.

Separately, I have also done some of the lazy-filter-wrapper work in a currently-private branch. Though in that case I did find that I was duplicating a lot of the `FilteredStrategy` code in the custom `FilteredSampleStrategy`.